cdbsearch.py

```diff
@@ -137,6 +131,52 @@ async def add_cdb_pv_positions(self, epd):
                 self.cdbPvToLeaf[board.epd()] = len(pv) - 1 - parsed
                 asyncio.ensure_future(self.queryall(board.epd()))
 
+    async def pv_has_proven_mate(self, epd, pv):
+        """check if the PV line is a proven mate on cdb, and if not help prove it"""
+        if not pv or pv[-1] != "checkmate":
+            return False
+        if pv == ["checkmate"]:
+            return True
+        board = chess.Board(epd)
+        if len(pv) % 2 == 0:  # we just need to check the defender's moves
+            board.push(chess.Move.from_uci(pv[0]))
+            return await self.pv_has_proven_mate(board.epd(), pv[1:])
+
+        scored_db_moves = await self.queryall(epd)
+        if len(scored_db_moves) - 1 != len(list(board.legal_moves)):
+            # there are unscored moves for epd
+            # help to construct a proof by querying all yet unscored moves of board
+            for move in board.legal_moves:
+                ucimove = move.uci()
+                if ucimove not in scored_db_moves:
+                    board.push(move)
+                    asyncio.ensure_future(self.queryall(board.epd()))
+                    self.count_unscored.inc()
+                    board.pop()
+            return False
+
+        # we need to check if the _given_ PV is a correct mating line:
+        for m in pv[:2]:
+            board.push(chess.Move.from_uci(m))
+        if not await self.pv_has_proven_mate(board.epd(), pv[2:]):
+            return False
+        for _ in [0, 1]:
+            board.pop()
+
+        for m in scored_db_moves:
```

Comment:
This for loop should be commented, it is not quite clear to me what it does.

Reply:
I had exactly the same thought. But then I thought it sits maybe nicer next to CHECKMATE, and if people use scripts or some such to data mine cdbsearch output they can always be sure it `eval` a numerical value.

Reply:
python allows _ to be used for variable whose value is irrelevant. frequently used in for loops where you just need to run a number of times

Reply:
So this loop will only be reached if all the moves have a score. So these are alternative defences for the to-be-mated side. We need to check that they are all valid mate PV's. We do this recursively, always checking just the defenders moves.

Reply:
suggestion for a comment:
```
# now we check if all the currently non-best moves also inevitably lead to the defender being mated
```